### DISCUSSION TITLE
How to block Window interaction when another Window should have focus

### DISCUSSION THREAD

**Discussion by danielgruethling** [microsoft/microsoft-ui-xaml#8850] [Q&A] [answered]
I have a main Window that should have its interaction blocked while another Window is spawned to show e.g. an error.
I can set the secondary Window to always be on top and have some obfuscation on the main Window but that is not the functionality I desire.

The FileOpenPicker has the functionality I want to recreate. When I open the FileOpenPicker and initialize it with my main Window it blocks all interaction with the main Window.

Is there a way for a normal Window to recreate this functionality?

**Comment by castorix** [comment]
Set it Modal (for example with [OverlappedPresenter.IsModal](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.overlappedpresenter.ismodal?view=windows-app-sdk-1.3))
 (or [ContentDialog](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.contentdialog?view=windows-app-sdk-1.3))
For an error message, you can use [Windows.UI.Popups.MessageDialog](https://learn.microsoft.com/en-us/uwp/api/windows.ui.popups.messagedialog?view=winrt-22621)

**Reply by danielgruethling** [reply]
This seems to be exactly what I was looking for but what I don't understand is why simply setting the property throws this exception.
![grafik](https://github.com/microsoft/microsoft-ui-xaml/assets/31628469/b5bee28a-82d1-4208-ae3a-c4c4447631fd)

The presenter is an OverlappedPresenter.
This happens when I try to change it from false to true.

ContentDialog makes it difficult to handle because two ContentDialogs can not be displayed at the same time. If I have a ContentDialog open to show some information or prompt for confirmation and during that time some error happens in a background worker I can not display the error or have to close the currently open ContentDialog.

MessageDialog would work but I can not alter its design.

**Comment by castorix** [comment, marked as answer]
> This seems to be exactly what I was looking for but what I don't understand is why simply setting the property throws this exception. ![grafik](https://user-images.githubusercontent.com/31628469/266611015-b5bee28a-82d1-4208-ae3a-c4c4447631fd.png)

Ah yes, I had made a test and I had to use GWL_HWNDPARENT too (not sure if it is the best method...) :
```

{
    Button btn1;
    Grid grid1 = new Grid()
    {
        Background = new SolidColorBrush(Microsoft.UI.Colors.Blue),
        HorizontalAlignment = HorizontalAlignment.Stretch,
        VerticalAlignment = VerticalAlignment.Stretch,
        Children =
        {
            (btn1 = new Button()
            {
                HorizontalAlignment = HorizontalAlignment.Center,
                VerticalAlignment = VerticalAlignment.Center,
                Content = "Click" }
            ),
        }
    };
    btn1.Click += (sender, e) =>
    {
        Console.Beep(5000, 10);
    };
    Window wnd1 = new Window()
    {
        Content = grid1
    };
    IntPtr hWnd1 = WinRT.Interop.WindowNative.GetWindowHandle(wnd1);
    Microsoft.UI.WindowId windowId1 = Microsoft.UI.Win32Interop.GetWindowIdFromWindow(hWnd1);
    Microsoft.UI.Windowing.AppWindow appWindow1 = Microsoft.UI.Windowing.AppWindow.GetFromWindowId(windowId1);
    appWindow1.Resize(new Windows.Graphics.SizeInt32(400, 300));
    appWindow1.Move(new Windows.Graphics.PointInt32(600, 400));
    Microsoft.UI.Windowing.OverlappedPresenter presenter = appWindow1.Presenter as Microsoft.UI.Windowing.OverlappedPresenter;

    IntPtr hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
    SetWindowLong(hWnd1, GWL_HWNDPARENT, hWnd);
    presenter.IsModal = true;

    wnd1.Closed += (sender, e) =>
    {   
        // Otherwise the focus on main window is lost...
        EnableWindow(hWnd, true);
    };

    wnd1.Activate();
    appWindow1.Show();            
}
```

With declarations :

       public const int GWL_STYLE = (-16);
       public const int GWL_EXSTYLE = (-20);
       public const int GWL_HWNDPARENT = (-8);
       public static IntPtr SetWindowLong(IntPtr hWnd, int nIndex, IntPtr dwNewLong)
       {
           if (IntPtr.Size == 4)
           {
               return SetWindowLongPtr32(hWnd, nIndex, dwNewLong);
           }
           return SetWindowLongPtr64(hWnd, nIndex, dwNewLong);
       }

       [DllImport("User32.dll", CharSet = CharSet.Auto, EntryPoint = "SetWindowLong")]
       public static extern IntPtr SetWindowLongPtr32(IntPtr hWnd, int nIndex, IntPtr dwNewLong);

       [DllImport("User32.dll", CharSet = CharSet.Auto, EntryPoint = "SetWindowLongPtr")]
       public static extern IntPtr SetWindowLongPtr64(IntPtr hWnd, int nIndex, IntPtr dwNewLong);

       [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
       public static extern bool EnableWindow(IntPtr hWnd, bool bEnable);

**Reply by ghord** [reply]
Note that I've had some issues when doing this (win app sdk 1.3). Namely ComboBox/ListView controls behave very strangely in a new window when parent window is disabled (you would need to disable parent window in modal scenario). It does not respond to clicks/mouse input normally, only every other time, seems bugged.
